File: Java/soln-backtracking-problems/permutations.java
/**
 @author Farheen Bano
  
 References-
 https://www.interviewbit.com/problems/permutations/
 https://leetcode.com/problems/permutations/
*/

import java.util.*;
import java.lang.*;
import java.io.*;

class Solution {
    public List<List<Integer>> permute(int[] nums) {        
        List<List<Integer>> result=new ArrayList<>();
        possiblePermutation(nums,new ArrayList<>(),result);
        return result;
    }
    
    public void possiblePermutation(int nums[], List<Integer> cur, List<List<Integer>> result){
        if(cur.size()==nums.length)
            result.add(new ArrayList<>(cur));
        
        else{
            for(int i=0;i<nums.length;i++){
                if(cur.contains(nums[i]))
                    continue;
                cur.add(nums[i]);
                possiblePermutation(nums,cur,result);
                cur.remove(cur.size()-1);
            }
        }
    }    
}

File: C++/soln-backtracking-problems/permutations.cpp
/**
 @author Farheen Bano
  
 Date 13-08-2021

 References-
 https://www.interviewbit.com/problems/permutations/
 https://leetcode.com/problems/permutations/
*/

class Solution {
public:
    vector<vector<int>> permute(vector<int>& nums) {
        vector<vector<int>> result;
        vector<int> cur;
        possiblePermutation(nums,cur,result);
        return result;
    }
    
    void possiblePermutation(vector<int>& nums, vector<int>& cur, vector<vector<int>>& result){        
        if(cur.size()==nums.size())
            result.push_back(cur);
                
        else{
            for(int i=0;i<nums.size();i++){
                if(find(cur.begin(), cur.end(), nums[i]) != cur.end())
                    continue;
                cur.push_back(nums[i]);
                possiblePermutation(nums,cur,result);
                cur.pop_back();
            }
        }
    }
};